### ISSUE TITLE
Plan on Implementation of Seraphis and Jamtis to Monero

### ISSUE THREAD

**Issue by One-horse-wagon** [seraphis-migration/wallet3#28]
Plan on Implementation of Seraphis and Jamtis to Monero
by one-horse-wagon, November 2, 2022

Seraphis and Jamtis will result in a huge change to the Monero eco-system.   Final development and inmplementation of Seraphis and Jamtis is an immense task which requires innovative thought and work.   Following is a basic plan proposal for beginning the process.

1.  Administrator. The core developers are the de facto Board of Directors for Monero.  It is vital for them to be 100% behind this project and to be ultimately responsible for it. Without their commitment, this project goes nowhere.

	They are to agree upon a full time administrator that will take charge of the entire process.  He shall serve, without pay, at their sole discretion.  This will mitigate any accusations of impropriety, and also allow easy dismissal of him, should the need arise.

	Administrator shall decide on the course of events to take place, what needs development, what CCS proposals should be made, and anything else that needs direction. He may consult with whomever he wishes, but will only answer to the core developers.

	An effective administrator is needed because he will have to deal with obstructionists, petty jealousies, tremendous egos, and prima donnas as the project unfolds.  If there is no strong administrator, decision and design by committee is only to invite factionism and failure.

2.  rbrunner7. I recommend rbrunner7 be officially chosen by the core developers as the Administrator.  He is very well informed on Monero, Seraphis, Jamtis, and widely known in the community.  He has been around for quite a while and looks to remain so.

3.  Assistant Administrator.  Upon assuming the position, the Administrator shall appoint at least one full time Assistant Administrator to help with the tasks at hand.  They will report only to him and also serve without pay, at his sole discretion.

	If the Administrator position should become vacant, the Assistant Administrator is in a position to help a new Administrator, appointed by the core developers, to transition into his duties.  This provides continuity under all circumstances.

4.  UkoeHB et al.  UkoeHB and others (to be named) are to have an immediate CCS proposal to write up exactly what needs to be done and in what sequence.  Priority is to be given to the simplest parts and pieces that can be publicly tested first.  Once implementation gets underway, changes and additions will be made by the administrator. A half way decent plan is better than no plan. 

5.  CCS proposals.  Administrator to take the steps from UkoeHB et al. sequential paper and encourage developers to file CCS proposals for the needed work. The Administrator shall authorize and monitor the progress of CCS proposals and be responsible for the dispensation of funds.  When necessary, Administrator shall have the authority to withdraw the CCS proposal from a recipient, with cause.    

6.  Test network. Immediate goal is to start a partial test network as soon as possible. The time frame is a few weeks, not months. Even if only one function of some kind is available, it is a start that should be immediately implemented and rigorously tested.  As more functions come to fruition, they are to be added, and tested in every possible way.

	The test network is to be open to the public from the beginning, with wide  encouragement to participate in it.  There are a number of advanced and experienced coders that quietly watch Monero from the sidelines.  A test network easy to find and get into will draw some of them to try it out. Greatly increasing the number of participants raises the probability of finding truly hidden bugs and vulnerabilities.   

7.  Code review.  Final code review to be done by three companies to eliminate errors.  

8. Full Release. The initial final release of Seraphis and Jamtis to take place after the code review.  
 
9.  New wallets for everyone.  This is going to be a major point of contention--asking the public to change their wallets and seed phrases. It will take several months to bring everyone on board.  It is expected the Monero community, such as found on Reddit, will step up to the task of smoothing out the concerns.  There is nothing extra that needs to be done here other than disseminating how-to information in the beginning.

10.  Marketing increased fungibility.  The possibility of Seraphis and Jamtis offering powerful increases in fungibility has wide appeal to the general public.  This theme should be pushed and marketed from the beginning.  It will more than offset any negative publicity from temporary vulnerabilities and bugs uncovered in the testing process.

**Comment by UkoeHB** [comment]
> New wallets for everyone. This is going to be a major point of contention--asking the public to change their wallets and seed phrases.

Users won't need new seed phrases, old wallets should upgrade seamlessly into new wallets (same seeds under the hood). Brand new wallets will get new seeds of course (Polyseed).

**Comment by j-berman** [comment]
To be clear, users would need new addresses to receive Monero to after the upgrade. See page 15: https://raw.githubusercontent.com/MoneroKon/meta/main/slides/2022/j-berman.pdf

(FWIW I tried to condense the major feature upgrades and UX improvements Seraphis/Jamtis would bring to make it as clear as possible for normal users in that Monerokon presentation. Still waiting on them to upload the video... any day now)

**Comment by ACK-J** [comment]
My .02 in regards to bullet point 7, as a professional penetration tester. 

I believe the review of randomx hired four different security consultancies instead of three. I would say three would be sufficient but I’d caution going any lower than that, and would recommend four. When scheduling the security audits avoid having them all at the same time as they will likely find redundant bugs. Instead, after each audit, fix the major issues discovered before subsequent audits. We should also provide the final reports of all previous audits to the current consultants, the more context you can give, the better. Lastly, Q4 will always be the busiest time of the year for security auditors, with Q1 being the easiest to book (and likely cheapest).

**Comment by rbrunner7** [comment]
> 4\. UkoeHB et al. UkoeHB and others (to be named) are to have an immediate CCS proposal to write up exactly what needs to be done and in what sequence.

I tried to come up with a list of modules (important C++ classes for the new wallet plus additional supporting code for each) and then think about a good order of implementation. Despite making a serious effort I did not get very far, for various reasons. In a recent meeting that we organized for starting with implementation work it even was difficult to identify merely two such modules, one for @DangerousFreedom1984 and one for @j-berman, that are reasonable starting points.

Personally I don't worry too much about this situation; in my long time in IT I was member of several projects that similarly built big and complicated things basically from zero, out on the green field, and initial "visibility" was usually terrible. Things in such projects become gradually visible over time, as work and experience and insights progress.

I am afraid we will have to endure a phase of uncertainty, where involved devs will have to make CCS submissions that basically read like "3 months more Seraphis and Jamtis wallet dev work".

**Comment by rbrunner7** [comment]
> 6\. Test network. Immediate goal is to start a partial test network as soon as possible. The time frame is a few weeks, not months. Even if only one function of some kind is available, it is a start that should be immediately implemented and rigorously tested. 

That's an excellent idea, and I will keep an eye on it to catch the earliest possible time to launch such a testnet.

However IMHO we should stay realistic: Such a testnet probably only starts to make sense if Seraphis transactions can go into the blockchain, and if we have a rudimentary wallet that can read them out from there, and make some simple new transactions that can go into the blockchain. For that to be possible, many components have to work already, and work together, and I am afraid that we are quite a number of months away from this point right now.

**Comment by rbrunner7** [comment]
> 10\. Marketing increased fungibility. The possibility of Seraphis and Jamtis offering powerful increases in fungibility has wide appeal to the general public. This theme should be pushed and marketed from the beginning.

Agreed, we should try that.

Monero and marketing in general is mostly a very mixed story, much marketing simply does not take place, or is even frowned upon by people who think that Monero should stand on its own and appeal by merit of its capabilities, but this situation is probably different.

I know the Monero subreddit very well and also have good "standing" there as a commenter / poster; I think it will be possible to reach many people by doing posts there, as one of the measures in this marketing push.

**Comment by rbrunner7** [comment]
With working group meetings taking place weekly now, and a first "modus operandi" established, I think we can close this issue as not under active discussion anymore. We will without doubt come back to one or another point that @One-horse-wagon rises here, which will then lead to new issue and/or discussions in the meetings.
